File: src/org/ligi/gobandroid/ui/GoBoardView.java
/**
 * gobandroid 
 * by Marcus -Ligi- Bueschleb 
 * http://ligi.de
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 3 as 
 * published by the Free Software Foundation; 
 * 
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details. 
 * 
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see <http://www.gnu.org/licenses/>.
 * 
 **/

package org.ligi.gobandroid.ui;

import org.ligi.gobandroid.logic.GoDefinitions;
import org.ligi.gobandroid.logic.GoGame;
import org.ligi.gobandroid.logic.GoMarker;
import org.ligi.gobandroid.logic.Logger;

import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.Canvas;
import android.graphics.Matrix;
import android.graphics.Paint;
import android.graphics.RectF;
import android.graphics.Paint.FontMetrics;
import android.util.Log;
import android.view.MotionEvent;
import android.view.View;


/**
 * Class to visually represent a Go Board in Android
 * 
 * @author <a href="http://ligi.de">Marcus -Ligi- Bueschleb</a>
 *         
 * This software is licensed with GPLv3 
 */
public class GoBoardView extends View {
   	private GoGame game;

	private Paint whitePaint;
	private Paint blackPaint;
	private Paint blackTextPaint;
	private Paint whiteTextPaint;
	private Paint boardPaint;
	private Paint gridPaint;
	private Paint gridPaint_h; // highlighted for cursor
    
    private Paint textPaint;
    
    private float stone_size;
    private float stone_size_zoomed;
    private float stone_size_normal;
    
    private float offset_x=0.0f;
    private float offset_y=0.0f;
    
    //TODO rename - name is now misleading
    public byte touch_x=-1;
    public byte touch_y=-1;

    //public boolean do_zoom=false;
    
    private Bitmap bg_bitmap=null;
    
    private Bitmap white_stone_bitmap=null;
    private Bitmap black_stone_bitmap=null;
    private Bitmap white_stone_bitmap_small=null;
    private Bitmap black_stone_bitmap_small=null;
    
    private boolean width_is_max;
    private boolean regenerate_stones_flag=true;
    
    
    public GoBoardView( Context context,GoGame game ) {
        super( context );
        this.game=game;
      //  this.board=game.getVisualBoard();
        
        whitePaint=new Paint();
        whitePaint.setColor(0xFFCCCCCC);
        whitePaint.setAntiAlias(true);
        
        whiteTextPaint=new Paint();
        whiteTextPaint.setColor(0xFFFFFFFF);
        whiteTextPaint.setAntiAlias(true);
        whiteTextPaint.setTextAlign(Paint.Align.CENTER );
        whiteTextPaint.setShadowLayer(2, 1, 1, 0xFF000000);
        
        blackTextPaint=new Paint();
        blackTextPaint.setColor(0xFF000000);
        blackTextPaint.setAntiAlias(true);
        blackTextPaint.setTextAlign(Paint.Align.CENTER  ); 
        blackTextPaint.setShadowLayer(2, 1, 1, 0xFFFFFFFF);
        
        blackPaint=new Paint();
        blackPaint.setColor(0xFF000000);
       
        blackPaint.setAntiAlias(true);
        
        boardPaint=new Paint();
        //boardPaint.setColor(0xFFc6b460);
        boardPaint.setColor(0xFFA77E3D);
        //boardPaint.setColor(0xFFA68064);
        gridPaint=new Paint();
        
        gridPaint_h=new Paint();
        gridPaint_h.setColor(0xFF0000FF);
        gridPaint_h.setShadowLayer(1,1,1,0xFFFFFFFF );
        //gridPaint.setColor(0xFFFFFFFF);
        //gridPaint.setShadowLayer(1,1,1,0xFF000000 );
        
        gridPaint.setColor(0xFF000000);
        gridPaint.setShadowLayer(1,1,1,0xFFFFFFFF );
        gridPaint.setTextAlign(Paint.Align.CENTER );
        gridPaint.setTextSize(12.0f );
    
        textPaint=new Paint();
        textPaint.setColor(0xFF000000);
        textPaint.setAntiAlias(false);
        
        setFocusable(true);   
    }


	public Bitmap resize_to_screen(Bitmap orig, float x_scale_, float y_scale_) {
		// createa matrix for the manipulation
		Matrix matrix = new Matrix();
		float x_scale, y_scale;
		if (y_scale_ != 0f)
			y_scale = (getHeight() * y_scale_) / orig.getHeight();
		else
			// take x_scale
			y_scale = (getWidth() * x_scale_) / orig.getWidth();

		if (x_scale_ != 0f)
			x_scale = (getWidth() * x_scale_) / orig.getWidth();
		else
			x_scale = (getHeight() * y_scale_) / orig.getHeight();

		matrix.postScale(x_scale, y_scale);
		return Bitmap.createBitmap(orig, 0, 0, (int) (orig.getWidth()),
				(int) (orig.getHeight()), matrix, true);// BitmapContfig.ARGB_8888
		// );
	}

    
    public void prepare_keyinput() {
    	if (touch_x==-1) touch_x=0;
    	if (touch_y==-1) touch_y=0;
    }
    

    public void setZoom(boolean zoom_flag) {
    	if (zoom_flag) 
    		{
    		stone_size=stone_size_zoomed;
			
			if (touch_x>= (2*game.getVisualBoard().getSize())/3)
				offset_x=-stone_size*game.getVisualBoard().getSize()+this.getWidth();
			else if (touch_x> (game.getVisualBoard().getSize()/3))
				offset_x=(-stone_size*game.getVisualBoard().getSize()+this.getWidth())/2;	
			
			if (touch_y>= (2*game.getVisualBoard().getSize())/3)
				offset_y=-stone_size*game.getVisualBoard().getSize()+this.getHeight();
			else if (touch_y> (game.getVisualBoard().getSize()/3))
				offset_y=(-stone_size*game.getVisualBoard().getSize()+this.getHeight())/2;	
    		}
    	else 
    		{
			stone_size=stone_size_normal;
			offset_x=0;
			offset_y=0;
    		}
    	
    	
    	regenerate_images();	
    	invalidate();
    }

    /**
     * check if the Board is Zoomed
     * 
     * @return true if zoomed
     */
    public boolean isZoomed() {
    	return (stone_size==stone_size_zoomed);
    }
    
    @Override
    protected void onDraw(Canvas canvas) {
    
    	
    	if (game==null)
    		return;
    	
    	if (regenerate_stones_flag)
    		regenerate_images();
    	
    	if (bg_bitmap!=null)
    		canvas.drawBitmap(bg_bitmap,0.0f, 0.0f, boardPaint);
    	else
    		canvas.drawRect(new RectF(0,0,this.getWidth(),this.getHeight()),boardPaint );
    	
    	
        canvas.translate(offset_x, offset_y);
        
        // draw the vertical lines
        for(int x=0;x<game.getVisualBoard().getSize();x++)
        	canvas.drawLine(stone_size/2.0f   + x*stone_size , stone_size/2.0f, stone_size/2.0f+ x*stone_size,stone_size*(float)(game.getVisualBoard().getSize()-1) +stone_size/2.0f,(touch_x==x)?gridPaint_h:gridPaint);	
        	
        // draw the horizontal lines and the legend
        for(int x=0;x<game.getVisualBoard().getSize();x++)
        {
            canvas.drawLine(stone_size/2.0f , stone_size/2.0f + x*stone_size , stone_size*(float)(game.getVisualBoard().getSize()-1)+stone_size/2.0f ,stone_size/2.0f+ x*stone_size, (touch_y==x)?gridPaint_h:gridPaint);
            
            if (GoPrefs.getLegendEnabled()) {
            	canvas.drawText("" + (1+x) , 6+ stone_size*(float)(game.getVisualBoard().getSize()-1)+stone_size/2.0f ,stone_size/2.0f+ x*stone_size+gridPaint.getTextSize()/3,gridPaint);
            	
            	if ((x>7)&&GoPrefs.getLegendSGFMode())
            		canvas.drawText("" + (char)('A'+(x+1)) , stone_size/2.0f+ x*stone_size,stone_size*(float)(game.getVisualBoard().getSize()-1) +stone_size/2.0f + 1 + gridPaint.getTextSize() ,gridPaint);
            	else
            		canvas.drawText("" + (char)('A'+x) , stone_size/2.0f+ x*stone_size,stone_size*(float)(game.getVisualBoard().getSize()-1) +stone_size/2.0f + 1 + gridPaint.getTextSize() ,gridPaint);
            }
        }
                
        
        
        for(byte x=0;x<game.getVisualBoard().getSize();x++)
            for(byte y=0;y<game.getVisualBoard().getSize();y++)
            {
            	blackPaint.setColor(0xFF000000);
            	blackPaint.setStrokeWidth(stone_size/12);
            	//blackPaint.setStyle(Paint.Style) .setStrokeWidth(stone_size/12);
            	
            	if (game.isPosHoschi(x, y))
            		canvas.drawCircle( stone_size/2.0f+ x*stone_size +0.5f ,stone_size/2.0f+y*stone_size+0.5f,stone_size/10,blackPaint );
            	
            	 
            	// paint the territory with alpha transparent stones
                if (game.isFinished()) { 

            		blackPaint.setColor(0x77000000);
            		whitePaint.setColor(0x77CCCCCC);
            		
                	if (game.area_assign[x][y]==GoDefinitions.PLAYER_BLACK)
                		canvas.drawBitmap(black_stone_bitmap, x*stone_size  ,y*stone_size,whitePaint );
                		
                        
                   	if (game.area_assign[x][y]==GoDefinitions.PLAYER_WHITE)
                		canvas.drawBitmap(white_stone_bitmap, x*stone_size  ,y*stone_size,whitePaint );
                   		
                			
                }
                
            	
            	
            	blackPaint.setColor(0xFF000000);
        		whitePaint.setColor(0xFFCCCCCC);
        		
            	if (game.getCalcBoard().isCellDead(x,y))
            	{
            		if (game.getVisualBoard().isCellWhite(x,y))
            			canvas.drawBitmap(white_stone_bitmap_small, x*stone_size  + (stone_size-white_stone_bitmap_small.getWidth())/2 ,y*stone_size + (stone_size-white_stone_bitmap_small.getHeight())/2,whitePaint );
            		
            		if (game.getVisualBoard().isCellBlack(x,y))
            			canvas.drawBitmap(black_stone_bitmap_small, x*stone_size  + (stone_size-black_stone_bitmap_small.getWidth())/2 ,y*stone_size + (stone_size-black_stone_bitmap_small.getHeight())/2,whitePaint );
            		
            	}
            	else
            	{
            		if (game.getVisualBoard().isCellWhite(x,y))
            			canvas.drawBitmap(white_stone_bitmap, x*stone_size  ,y*stone_size,whitePaint );
            		if (game.getVisualBoard().isCellBlack(x,y))
            			canvas.drawBitmap(black_stone_bitmap, x*stone_size  ,y*stone_size,whitePaint );
            
            		if (GoPrefs.getMarkLastStone()) { // if the last stone should be marked
            			blackPaint.setStyle(Paint.Style.STROKE);
            			whitePaint.setStyle(Paint.Style.STROKE);
            			whitePaint.setStrokeWidth(2.0f);
            			blackPaint.setStrokeWidth(2.0f);
            		
            			if ((game.getActMove().getX()==x)&&(game.getActMove().getY()==y))
            			{
            				if (game.getVisualBoard().isCellWhite(x,y))
            					canvas.drawCircle( stone_size/2.0f+ x*stone_size  ,stone_size/2.0f+y*stone_size,stone_size/4.0f,blackPaint );
                			if (game.getVisualBoard().isCellBlack(x,y))
            					canvas.drawCircle( stone_size/2.0f+ x*stone_size  ,stone_size/2.0f+y*stone_size,stone_size/4.0f,whitePaint );
            	
            			}
            		}
            	}

            	
            }

        FontMetrics fm=whiteTextPaint.getFontMetrics();
        
        // paint the markers
        for (GoMarker marker:game.getActMove().getMarkers())
        	        
        if (game.getVisualBoard().isCellBlack(marker.getX(),marker.getY()))
    		canvas.drawText( marker.getText() , marker.getX()*stone_size + stone_size/2.0f ,marker.getY()*stone_size-(fm.top+fm.bottom) ,whiteTextPaint );
    	else
    		canvas.drawText( marker.getText() , marker.getX()*stone_size + stone_size/2.0f ,marker.getY()*stone_size-(fm.top+fm.bottom) ,blackTextPaint );
        
        
        canvas.restore();
    } // end of onDraw
    
    
    
    public void regenerate_images() {
   
    	try {

    	if (game==null)
    		return;
    	    	
    	width_is_max=(this.getWidth()<=this.getHeight());
       
        if (this.getWidth()<=this.getHeight())
            stone_size_normal=this.getWidth()/(float)game.getVisualBoard().getSize();
        else
            stone_size_normal=this.getHeight()/(float)game.getVisualBoard().getSize();
       
        stone_size=stone_size_normal;
        stone_size_zoomed=stone_size_normal*2;

        
    	Log.i("gobandrod","regenerating images to stone size " + stone_size + "  " +regenerate_stones_flag); 
    	float SMALL_STONE_SCALER=0.6f;	
    	white_stone_bitmap=GOSkin.getWhiteStone(stone_size);
    	black_stone_bitmap=GOSkin.getBlackStone(stone_size);
    	white_stone_bitmap_small=GOSkin.getWhiteStone(SMALL_STONE_SCALER*stone_size);
    	black_stone_bitmap_small=GOSkin.getBlackStone(SMALL_STONE_SCALER*stone_size);
    
    	bg_bitmap=GOSkin.getBoard(this.getWidth(),this.getHeight());
    	regenerate_stones_flag=false;
    	
    	whiteTextPaint.setTextSize(stone_size);
    	blackTextPaint.setTextSize(stone_size);
    	
    	invalidate();
    	}
    	catch (Exception e) {
    		
    		Logger.i("Exception " + e);
    	}
    }	
    
    @Override
    protected void onSizeChanged(int w, int h, int oldw, int oldh) {
    	regenerate_images();        
    }


    public void doTouch( MotionEvent event) {
    	
    	if (game==null)
    		return;
    	
    	float virtualTouchX=event.getX()-offset_x;
    	float virtualTouchY=event.getY()-offset_y;
    	
    	
    	float board_size=stone_size*game.getVisualBoard().getSize();

    	if ((virtualTouchY<board_size)&&(virtualTouchX<board_size)) // if user put his finger on the board
        {
        	touch_x=(byte)(virtualTouchX/stone_size);
    		touch_y=(byte)(virtualTouchY/stone_size);
    		if (event.getAction()==MotionEvent.ACTION_UP)
    		{
    		if (isZoomed()||(!GoPrefs.getFatFingerEnabled()))
    		{
    			if (!game.do_move(touch_x,touch_y))	;
        		touch_x=-1;
        		touch_y=-1;
        		
        		setZoom(false);
        		stone_size=stone_size_normal;
        		
    		}
    		else
    			setZoom(true);
    		}
        }
        invalidate();  // the board looks diffrent after a move
     }

    public void setGame(GoGame game) {
    	this.game=game;
    	Logger.i("set game" + game);
    	regenerate_images();
    	
    }
}

File: android/src/main/java/org/ligi/gobandroid_hd/ui/GoBoardView.kt
/**
 * gobandroid
 * by Marcus -Ligi- Bueschleb
 * http://ligi.de
 *
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 3 as
 * published by the Free Software Foundation;
 *
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 *
 * You should have received a copy of the GNU General Public License
 * along with this program. If not, see //www.gnu.org/licenses/>.
 */

package org.ligi.gobandroid_hd.ui

import android.content.Context
import android.graphics.*
import android.util.AttributeSet
import android.view.View
import com.github.salomonbrys.kodein.instance
import com.github.salomonbrys.kodein.lazy
import org.ligi.gobandroid_hd.App
import org.ligi.gobandroid_hd.InteractionScope
import org.ligi.gobandroid_hd.R
import org.ligi.gobandroid_hd.logic.Cell
import org.ligi.gobandroid_hd.logic.CellImpl
import org.ligi.gobandroid_hd.logic.GoDefinitions
import org.ligi.gobandroid_hd.logic.GoGame
import org.ligi.gobandroid_hd.model.GameProvider
import timber.log.Timber
import java.io.File
import java.io.FileOutputStream

/**
 * Class to visually represent a Go Board in Android
 */
open class GoBoardView : View {

    constructor(context: Context, attrs: AttributeSet) : super(context, attrs)

    constructor(context: Context) : super(context)

    private val SMALL_STONE_SCALE_FACTOR = 0.6f

    val gameProvider: GameProvider  by App.kodein.lazy.instance()
    val interactionScope: InteractionScope by App.kodein.lazy.instance()

    private var zoom_poi: Cell? = null

    var do_legend = true
    var do_actpos_highlight = true
    var do_actpos_highlight_ony_if_active = true
    var mark_last_stone = true
    var legend_sgf_mode = true // GoPrefs.getLegendSGFMode()

    // we need a lot of paints, but so we have a efficient onDraw with less
    // calls and the mem does not matter compared to bitmaps
    private val hoshi_paint = Paint()
    private val legendPaint = Paint()
    private val blackTextPaint by lazy {
        val res = Paint(whiteTextPaint)
        res.color = Color.BLACK
        res.setShadowLayer(2f, 1f, 1f, Color.WHITE)
        res
    }
    private val whiteTextPaint = Paint()
    private val gridPaint = Paint()
    private val gridPaint_h = Paint()
    private val bitmapPaint = Paint()
    private val placeStonePaint = Paint()
    private val opaque_paint = Paint()
    private val whiteLastStoneCirclePaint = Paint()
    private val blackLastStoneCirclePaint by lazy {
        Paint(whiteLastStoneCirclePaint).apply {
            color = Color.BLACK
        }
    }

    private var stone_size: Float = 0.toFloat()

    private var white_stone_bitmap: Bitmap? = null
    private var black_stone_bitmap: Bitmap? = null
    private var white_stone_bitmap_small: Bitmap? = null
    private var black_stone_bitmap_small: Bitmap? = null

    var move_stone_mode = false

    private var regenerate_stones_flag = true

    protected open fun enforceSquare() = true

    var zoom = 1.0f
        /**
         * set the zoom factor - 1.0 ( default ) means no zoom
         */
        set(value) {
            field = value
            regenerateStoneImagesWithNewSize()
        }


    init {
        init()
    }

    fun init() {

        // these paint init's should be exposed to a designer ^^

        // paint to make the stones transparent in game-result mode ( territory
        // stones ) - also used for stone move and shadow stone
        opaque_paint.alpha = 0x77

        // for marking the last stone
        whiteLastStoneCirclePaint.color = Color.WHITE
        whiteLastStoneCirclePaint.style = Paint.Style.STROKE
        whiteLastStoneCirclePaint.strokeWidth = 2.0f
        whiteLastStoneCirclePaint.isAntiAlias = true

        // for the star or hoshi points
        hoshi_paint.color = Color.BLACK
        hoshi_paint.style = Paint.Style.FILL
        hoshi_paint.isAntiAlias = true

        // for drawing on the stones
        whiteTextPaint.color = Color.WHITE
        whiteTextPaint.isAntiAlias = true
        whiteTextPaint.textAlign = Paint.Align.CENTER
        whiteTextPaint.setShadowLayer(2f, 1f, 1f, Color.BLACK)

        gridPaint_h.color = 0xFF0000FF.toInt()

        gridPaint.color = Color.BLACK

        legendPaint.color = Color.BLACK
        legendPaint.textAlign = Paint.Align.CENTER
        legendPaint.textSize = 12.0f
        legendPaint.isAntiAlias = true

        placeStonePaint.alpha = 127

        // defaults
        isFocusable = true
    }

    val game: GoGame
        get() = gameProvider.get()


    val zoomTranslate: PointF
        get() {
            if (zoom <= 1.0f) {
                return PointF(0f, 0f)
            }

            val act_zoom_point = calcActZoomPOI() ?: return PointF(0f, 0f)

            return PointF(-stone_size * (act_zoom_point.x - game.size.toFloat() / 2.0f / zoom),
                    -stone_size * (act_zoom_point.y - game.size.toFloat() / 2.0f / zoom))
        }

    private fun calcActZoomPOI(): Cell? {
        return if (zoom_poi != null) {
            zoom_poi
        } else if (interactionScope.touchCell != null) {
            interactionScope.touchCell
        } else {
            Timber.w("zoom requested but no POI to center around")
            return null
        }
    }

    /**
     * redraw board when draw_board is ture
     */
    fun screenshot(file: File, bitmap: Bitmap) {

        draw(Canvas(bitmap))

        try {
            file.createNewFile()
            val out = FileOutputStream(file)
            bitmap.compress(Bitmap.CompressFormat.PNG, 90, out)
            out.close()
        } catch (e: Exception) {
            e.printStackTrace()
        }

    }

    override fun onDraw(canvas: Canvas) {
        canvas.save()

        // when we have zoomed in - center translate the canvas around the POI
        if (zoom > 1.0f) {
            canvas.translate(zoomTranslate.x, zoomTranslate.y)
        }

        if (regenerate_stones_flag) {
            regenerate_images()
        }

        var actpos_highlight_condition = false

        if (!(do_actpos_highlight_ony_if_active && !isFocused)) {
            actpos_highlight_condition = do_actpos_highlight && interactionScope.hasTouchCell()
        }

        // draw semi transparent stone on current touch pos as a shadow
        if (!move_stone_mode && actpos_highlight_condition) {
            val touch_cell = interactionScope.touchCell
            if (touch_cell != null) {
                val bitmap = if (game.isBlackToMove) black_stone_bitmap!! else white_stone_bitmap!!
                canvas.drawBitmap(bitmap, touch_cell.x * stone_size, touch_cell.y * stone_size, placeStonePaint)
            }
        }

        // draw the vertical lines for the grid
        val board = game.visualBoard
        for (x in 0..game.size - 1)
            canvas.drawLine(stone_size / 2.0f + x * stone_size,
                    stone_size / 2.0f,
                    stone_size / 2.0f + x * stone_size,
                    stone_size * (board.size - 1).toFloat() + stone_size / 2.0f,
                    if (actpos_highlight_condition && interactionScope.touchCell!!.x == x) gridPaint_h else gridPaint)

        // draw the horizontal lines and the legend
        for (x in 0..board.size - 1) {
            canvas.drawLine(stone_size / 2.0f,
                    stone_size / 2.0f + x * stone_size,
                    stone_size * (board.size - 1).toFloat() + stone_size / 2.0f,
                    stone_size / 2.0f + x * stone_size,
                    if (actpos_highlight_condition && interactionScope.touchCell!!.y == x) gridPaint_h else gridPaint)
            if (do_legend) {
                canvas.drawText("" + (game.size - x), legendPaint.textSize / 2f + stone_size * (game.size - 1).toFloat() + stone_size / 2.0f,
                        stone_size / 2.0f + x * stone_size + gridPaint.textSize / 3,
                        legendPaint)

                canvas.drawText(getLegendLetter(x),
                        stone_size / 2.0f + x * stone_size,
                        stone_size * (game.size - 1).toFloat() + stone_size / 2.0f + 1f + legendPaint.textSize,
                        legendPaint)
            }
        }


        board.statelessGoBoard.withAllCells { cell ->
            if (game.isCellHoshi(cell)) {
                drawBoardCircle(canvas, cell.x.toFloat(), cell.y.toFloat(), 2f + stone_size / 10, hoshi_paint)
            }

            // paint the territory with alpha opaque stones
            if (game.scorer != null) {
                if (game.scorer!!.area_assign[cell.x][cell.y] == GoDefinitions.PLAYER_BLACK)
                    canvas.drawBitmap(black_stone_bitmap!!, cell.x * stone_size, cell.y * stone_size, opaque_paint)
                else if (game.scorer!!.area_assign[cell.x][cell.y] == GoDefinitions.PLAYER_WHITE)
                    canvas.drawBitmap(white_stone_bitmap!!, cell.x * stone_size, cell.y * stone_size, opaque_paint)

            }

            if (board.isCellDeadWhite(cell)) {
                canvas.drawBitmap(white_stone_bitmap_small!!,
                        cell.x * stone_size + (stone_size - white_stone_bitmap_small!!.width) / 2,
                        cell.y * stone_size + (stone_size - white_stone_bitmap_small!!.height) / 2,
                        bitmapPaint)
            } else if (board.isCellDeadBlack(cell)) {
                canvas.drawBitmap(black_stone_bitmap_small!!,
                        cell.x * stone_size + (stone_size - black_stone_bitmap_small!!.width) / 2,
                        cell.y * stone_size + (stone_size - black_stone_bitmap_small!!.height) / 2,
                        bitmapPaint)
            } else if (board.isCellWhite(cell)) {
                canvas.drawBitmap(white_stone_bitmap!!, cell.x * stone_size, cell.y * stone_size, getStonePaintForCell(cell))
            } else if (board.isCellBlack(cell)) {
                canvas.drawBitmap(black_stone_bitmap!!, cell.x * stone_size, cell.y * stone_size, getStonePaintForCell(cell))
            }

        }

        if (mark_last_stone) {
            val lastMoveCell = game.actMove.cell
            if (lastMoveCell != null) {
                val paint = if (board.isCellWhite(lastMoveCell)) blackLastStoneCirclePaint else whiteLastStoneCirclePaint
                drawBoardCircle(canvas, lastMoveCell.x.toFloat(), lastMoveCell.y.toFloat(), 2f + stone_size / 4f, paint)
            }
        }

        // paint the markers
        for (marker in game.actMove.markers) {
            val markerPaint = getTextPaintForCell(marker)
            val x = marker.x * stone_size + stone_size / 2.0f
            val y = marker.y * stone_size + stone_size / 2.0f
            marker.draw(canvas, stone_size, x, y, markerPaint)
        }

        canvas.restore()
    }

    /**
     * used to make nice code around hoshi and last stone circle
     */
    private fun drawBoardCircle(canvas: Canvas, x: Float, y: Float, size: Float, paint: Paint) {
        canvas.drawCircle(stone_size / 2.0f + x * stone_size, stone_size / 2.0f + y * stone_size, size, paint)
    }

    private fun getLegendLetter(x: Int): String {
        if (x > 7 && legend_sgf_mode) {
            return ('A' + (x + 1)).toString()
        }
        return ('A' + x).toString()
    }

    private fun getStonePaintForCell(cell: Cell): Paint {
        if (move_stone_mode && cell == game.actMove.cell) {
            return opaque_paint
        } else {
            return bitmapPaint
        }
    }

    private fun getTextPaintForCell(cell: Cell): Paint {
        if (game.visualBoard.isCellBlack(cell)) {
            return whiteTextPaint
        } else {
            return blackTextPaint
        }

    }

    private fun getScaledRes(size: Float, resID: Int): Bitmap {
        val unscaled_bitmap = BitmapFactory.decodeResource(this.resources, resID)
        return Bitmap.createScaledBitmap(unscaled_bitmap, size.toInt(), size.toInt(), true)
    }

    /**
     * resize the images regarding to stone_size
     */
    private fun regenerate_images() {

        white_stone_bitmap = getScaledRes(stone_size, R.drawable.stone_white)
        black_stone_bitmap = getScaledRes(stone_size, R.drawable.stone_black)
        white_stone_bitmap_small = getScaledRes(stone_size * SMALL_STONE_SCALE_FACTOR, R.drawable.stone_white)
        black_stone_bitmap_small = getScaledRes(stone_size * SMALL_STONE_SCALE_FACTOR, R.drawable.stone_black)

        regenerate_stones_flag = false

        whiteTextPaint.textSize = stone_size
        blackTextPaint.textSize = stone_size

    }

    fun setLineSize(size: Float) {
        gridPaint.strokeWidth = size
        gridPaint_h.strokeWidth = size
    }

    override fun onSizeChanged(w: Int, h: Int, oldw: Int, oldh: Int) {
        setSize(w, h)
    }

    fun setSize(w: Int, h: Int) {
        stone_size = zoom * (Math.min(w, h) / game.size.toFloat())
        legendPaint.textSize = stone_size / 4
        regenerate_stones_flag = true
    }

    fun regenerateStoneImagesWithNewSize() {
        setSize(width, height)
    }

    override fun onMeasure(widthMeasureSpec: Int, heightMeasureSpec: Int) {
        super.onMeasure(widthMeasureSpec, heightMeasureSpec)

        if (enforceSquare()) {
            val parentWidth = View.MeasureSpec.getSize(widthMeasureSpec)
            val parentHeight = View.MeasureSpec.getSize(heightMeasureSpec)
            val size = Math.min(parentWidth, parentHeight)
            setMeasuredDimension(size, size)
        }
    }

    fun setZoomPOI(zoom_poi: Cell) {
        this.zoom_poi = zoom_poi
        // TODO check use-cases if we need to invalidate here
    }

    fun pixel2cell(pixelX: Float, pixelY: Float): Cell? {
        val cellX = ((pixelX - zoomTranslate.x) / stone_size).toInt()
        val cellY = ((pixelY - zoomTranslate.y) / stone_size).toInt()
        val cell = CellImpl(cellX, cellY)

        if (game.calcBoard.isCellOnBoard(cell)) {
            return cell
        }
        return null
    }

}